### PR TITLE
feat(event_sources): add Secrets Manager secret rotation event

### LINKED ISSUE #2855: Feature request: Secrets Events
### Use case

When setting up secrets rotation you can rotate the secret via a specific lambda, therefore I think having a secrets event would be useful here.

### Solution/User Experience

Setting up a secrets event, similar to the way the other events are handled parsing the json object into a dataclass.

### Alternative solutions

```markdown
I'd be happy to help codify this if someone where to help outline the basic requirements.
```

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [X] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by roger-zhangg**
**Issue number:** #2855

### Changes

> Please provide a summary of what's being changed

Added a new data class `SecretManagerEvent` for parsing [event](https://docs.aws.amazon.com/secretsmanager/latest/userguide/rotating-secrets.html) from Secret Manager

### User experience

> Please share what the user experience looks like before and after this change

before
```python
def lambda_handler(event, context):
    arn = event['SecretId']
    token = event['ClientRequestToken']
    step = event['Step']
    ...
```

now
```python
from aws_lambda_powertools.utilities import parameters
from aws_lambda_powertools.utilities.data_classes import SecretsManagerEvent, event_source

secrets_provider = parameters.SecretsProvider()


@event_source(data_class=SecretsManagerEvent)
def lambda_handler(event: SecretsManagerEvent, context):
    # Getting secret value using Parameter utility
    # See https://docs.powertools.aws.dev/lambda/python/latest/utilities/parameters/
    secret = secrets_provider.get(event.secret_id, VersionId=event.client_request_token, VersionStage="AWSCURRENT")

    # You need to work with secrets afterwards
    # Check more examples: https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas

    return secret
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by roger-zhangg** [comment]
Currently this PR only included basic Data class support, but there are some other stuff we can do to provide a better UX.
- possible UX improvement - > add a `describe_secret` func
```python
# past
metadata = service_client.describe_secret(SecretId=event['SecretId'])

# possible
metadata = event.describe_secret(client = service_client)
```
We can also build a data class for `metadata` if needed.
See - https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/secretsmanager/client/describe_secret.html

- possible UX improvement - > add a `get_secret_value` func
```python
# past
secret = service_client.get_secret_value(SecretId=event['SecretId'], VersionId=event['ClientRequestToken'], VersionStage=stage)

# possible
secret = event.get_secret_value(client = service_client, VersionStage=stage)
```
We can also build a data class for `secret` if needed.
See - https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/secretsmanager/client/get_secret_value.html
```

**Comment by leandrodamascena** [comment]
> Currently this PR only included basic Data class support, but there are some other stuff we can do to provide a better UX.

Hi @roger-zhangg! I like helping customers with utilities that make their lives easier. However, we already have the [Parameter](https://docs.powertools.aws.dev/lambda/python/latest/utilities/parameters/) utility and customers can retrieve secrets/ssm/appconfig/dynamodb values using this utility.

For now, we will keep this as simple as possible by providing the Data Class Event Source experience.

File: aws_lambda_powertools/utilities/data_classes/__init__.py [hunk 1]
```diff
@@ -26,6 +27,7 @@
 __all__ = [
     "APIGatewayProxyEvent",
     "APIGatewayProxyEventV2",
+    "SecretManagerEvent",
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/__init__.py]:
```suggestion
    "SecretsManagerEvent",
```

File: aws_lambda_powertools/utilities/data_classes/secrets_event.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from typing_extensions import Literal
+
+from aws_lambda_powertools.utilities.data_classes.common import DictWrapper
+
+
+class SecretManagerEvent(DictWrapper):
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/secrets_event.py]:
```suggestion
class SecretsManagerEvent(DictWrapper):
```

File: tests/unit/data_classes/test_secret_event.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from aws_lambda_powertools.utilities.data_classes.secrets_event import SecretManagerEvent
+from tests.functional.utils import load_event
+
+
+def test_vpc_lattice_event():
```
Comment (leandrodamascena) [inline tests/unit/data_classes/test_secret_event.py]:
```suggestion
def test_secrets_manager_event():
```

File: aws_lambda_powertools/utilities/data_classes/secrets_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/secrets_event.py:1]: please rename this file to `secrets_manager_event.py`

File: tests/unit/data_classes/test_secret_event.py
Comment (leandrodamascena) [inline tests/unit/data_classes/test_secret_event.py:1]: please rename this file to `test_secrets_manager_event.py`

File: tests/unit/data_classes/test_secret_event.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    raw_event = load_event("secretManagerEvent.json")
```
Comment (leandrodamascena) [inline tests/unit/data_classes/test_secret_event.py]:
```suggestion
    raw_event = load_event("secretsManagerEvent.json")
```

File: tests/events/secretManagerEvent.json
Comment (leandrodamascena) [inline tests/events/secretManagerEvent.json:1]: please rename this file to  `secretsManagerEvent.json`

File: aws_lambda_powertools/utilities/data_classes/__init__.py (hunk 1 above)
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/__init__.py]:
```suggestion
    "SecretsManagerEvent",
```

File: docs/utilities/data_classes.md [hunk 5]
```diff
@@ -1161,6 +1161,26 @@ You can register your Lambda functions as targets within an Amazon VPC Lattice s
...
+
+    ```python
+    from aws_lambda_powertools.utilities.data_classes import event_source, SecretManagerEvent
+
+    @event_source(data_class=SecretManagerEvent)
+    def lambda_handler(event: SecretManagerEvent, context):
+        # Multiple records can be delivered in a single event
+        service_client = boto3.client('secretsmanager', endpoint_url=os.environ['SECRETS_MANAGER_ENDPOINT'])
+        create_secret(service_client, event.secret_id, event.client_request_token)
+        ...
```
Comment (leandrodamascena) [inline docs/utilities/data_classes.md]: What do you think about making this example more realistic? It's missing the boto3 import, the `create_secret` function is missing, and the comment is also wrong. Can you improve a little?

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @roger-zhangg! I made some comments to improve this PR!

You are rocking it! :rocket: :star2:

File: docs/utilities/data_classes.md (hunk 5 above)
Comment (roger-zhangg) [reply docs/utilities/data_classes.md]: this is a example from https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas, but they seems too complicated for a simple example. I extracted `service_client.get_secret_value` to demo our case here. Please help to check if this looks good to you. Thanks!

Comment (leandrodamascena) [reply docs/utilities/data_classes.md]: Hey @roger-zhangg! I changed some lines in the example. Please check it and see if you agree.

Comment (roger-zhangg) [reply docs/utilities/data_classes.md]: Thanks for the fix, I'm not familiar to parameter so thanks for pointing it out!

**Review by leandrodamascena — DISMISSED** [review]
Approved!

**Comment by heitorlessa** [comment]
Reviewing today, please update PR description UX so it's always up to date. `create_secret` isn't accurate.

File: aws_lambda_powertools/utilities/data_classes/secrets_manager_event.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+class SecretsManagerEvent(DictWrapper):
+    @property
+    def secret_id(self) -> str:
+        """SecretId: The secret ARN or identifier"""
+        return self["SecretId"]
+
+    @property
+    def client_request_token(self) -> str:
+        """ClientRequestToken: The ClientRequestToken of the secret version"""
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/secrets_manager_event.py]:
```suggestion
        """ClientRequestToken: The ClientRequestToken associated with the secret version"""
```

File: aws_lambda_powertools/utilities/data_classes/secrets_manager_event.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        return self["ClientRequestToken"]
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/secrets_manager_event.py]:
```suggestion
    @property
    def client_request_token(self) -> str:
        """ClientRequestToken: The ClientRequestToken of the secret version"""
        return self["ClientRequestToken"]
        
    @property
    def version_id(self) -> str:
        """Alias to ClientRequestToken to get token associated to version"""
        return self["ClientRequestToken"]        
```

the UX is quite confusing due to API interchangeably using `ClientRequestToken` and `VersionId` with the same value in different API calls. We can make it easier for customers (cognitive load) by having an alias here to what the intent actually is.

File: examples/event_sources/src/secrets_manager.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+from aws_lambda_powertools.utilities.data_classes import SecretsManagerEvent, event_source
+
+secrets_provider = parameters.SecretsProvider()
+
+
+@event_source(data_class=SecretsManagerEvent)
+def lambda_handler(event: SecretsManagerEvent, context):
+    # Getting secret value using Parameter utility
+    # See https://docs.powertools.aws.dev/lambda/python/latest/utilities/parameters/
+    secret = secrets_provider.get(event.secret_id, VersionId=event.client_request_token, VersionStage="AWSCURRENT")
```
Comment (heitorlessa) [inline examples/event_sources/src/secrets_manager.py]:
```suggestion
    secret = secrets_provider.get(event.secret_id, VersionId=event.version_id, VersionStage="AWSCURRENT")
```

File: tests/unit/data_classes/test_secrets_manager_event.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
+from aws_lambda_powertools.utilities.data_classes.secrets_manager_event import SecretsManagerEvent
+from tests.functional.utils import load_event
+
+
+def test_secrets_manager_event():
+    raw_event = load_event("secretsManagerEvent.json")
+    parsed_event = SecretsManagerEvent(raw_event)
+
+    assert parsed_event.secret_id == raw_event["SecretId"]
+    assert parsed_event.client_request_token == raw_event["ClientRequestToken"]
```
Comment (heitorlessa) [inline tests/unit/data_classes/test_secrets_manager_event.py:10]:
```suggestion
    assert parsed_event.version_id == raw_event["ClientRequestToken"]
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
tiny changes to improve UX :-)

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**